dev/tox.ini

```diff
@@ -18,4 +18,4 @@ ignore=E226,E241,E305,E402,E722,E731,E741,W503,W504
 max-line-length=100
 exclude=cloudpickle.py,heapq3.py,shared.py,python/docs/conf.py,work/*/*.py,python/.eggs/*,dist/*,.git/*
 [pydocstyle]
-ignore=D100,D101,D102,D103,D104,D105,D106,D107,D200,D201,D202,D203,D204,D205,D206,D207,D208,D209,D210,D211,D212,D213,D214,D215,D300,D301,D302,D400,D401,D402,D403,D404,D405,D406,D407,D408,D409,D410,D411,D412,D413,D414
+ignore=D100,D101,D102,D103,D104,D105,D106,D107,D200,D201,D202,D203,D204,D205,D206,D207,D208,D209,D210,D211,D212,D213,D214,D215,D300,D301,D302,D400,D401,D402,D403,D404,D405,D406,D407,D408,D409,D410,D411,D412,D413,D414,D415,D416,D417
```

Comment:
Out of curiosity what are the additional linter rules that are ignored here?

Reply:
I had to add these new ignore rules because, presumably, when pydocstyle was last run it was at an older version that didn't implement these checks.

Here's an example of each:

```
./sql/gen-sql-api-docs.py:121 in private function `_make_pretty_note`:
        D415: First line should end with a period, question mark, or exclamation point (not 't')
./sql/gen-sql-api-docs.py:163 in public function `generate_sql_markdown`:
        D416: Section name should end with a colon ('Arguments:', not 'ARGUMENTS')
./sql/gen-sql-api-docs.py:163 in public function `generate_sql_markdown`:
        D417: Missing argument descriptions in the docstring (argument(s) jvm, path are missing descriptions in 'generate_sql_markdown' docstring)
```

Honestly, the list of disabled pydocstyle tests is so long I wonder why we have the thing at all! 😆

Perhaps in a separate PR we can start fixing some of these issues and reenable the various checks. But I imagine the changes will touch every docstring we have.

Reply:
I mean, it was entirely disabled right? maybe for a reason. But a little is probably a bit better than nothing.

Reply:
Yep, totally disabled on both Jenkins and GitHub. I think it was unintentional, because `lint-python` accepts when pydocstyle is missing and just happily carries along.

I'm honestly skeptical, given the lengthy list of ignore codes, that we are checking _anything_ with pydocstyle at the moment, even with this fix. In fact, looking at the [reference list of error codes](http://www.pydocstyle.org/en/5.0.2/error_codes.html), I believe that's the case!

@shaneknapp / @HyukjinKwon - Are you aware that we haven't been running pydocstyle tests for some time now? Was it intentionally disabled at some point? Even with this PR re-enabling it, I believe we are not actually testing anything, because we've configured the tool to ignore every error it checks for.

Reply:
I don't think it's intentional. Actually, I have been considering removing `pydocstyle` out which doesn't verify things very useful.

Reply:
Also, I think we didn't mention that we comply [PEP-257](ttps://www.python.org/dev/peps/pep-0257/) yet anywhere. Of course, It's good to incrementally fix to comply PEP-257; however, it ignores too many rules currently. I wonder if we added this too early without a comprehensive investigation. There were several rules that looked counterintuitive to me, IIRC too. cc @BryanCutler who, IIRC, also pointed this out.

BTW, many rules were disabled due to our legacy Epydoc, which I managed to remove at https://github.com/apache/spark/commit/fe75ff8bea3330a10aba1a61f3aba42e541195a8. So, several rules might be able to enable back now.